src/loader/scene.rs

```diff
@@ -0,0 +1,217 @@
+}
+
+impl FromWorld for super::ObjLoader {
+    fn from_world(world: &mut World) -> Self {
+        let supported_compressed_formats = match world.get_resource::<RenderDevice>() {
+            Some(render_device) => CompressedImageFormats::from_features(render_device.features()),
+            None => CompressedImageFormats::all(),
+        };
+        Self {
+            supported_compressed_formats,
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum ObjError {
+    #[error("Invalid OBJ file: {0}")]
+    TobjError(#[from] tobj::LoadError),
+    #[error("Failed to load materials for {0}: {1}")]
+    MaterialError(PathBuf, #[source] tobj::LoadError),
+    #[error("Invalid image file for texture: {0}")]
+    InvalidImageFile(PathBuf),
+    #[error("Asset reading failed: {0}")]
+    AssetIOError(#[from] bevy_asset::AssetIoError),
+    #[error("Texture conversion failed: {0}")]
+    TextureError(#[from] bevy_render::texture::TextureError),
+}
+
+pub(super) async fn load_obj<'a, 'b>(
+    bytes: &'a [u8],
+    load_context: &'a mut LoadContext<'b>,
+    supported_compressed_formats: CompressedImageFormats,
+) -> Result<(), ObjError> {
+    let obj = load_obj_scene(bytes, load_context, supported_compressed_formats).await?;
+    load_context.set_default_asset(LoadedAsset::new(obj));
+    Ok(())
+}
+
+async fn load_texture_image<'a, 'b>(
+    image_path: &'a str,
+    load_context: &'a mut LoadContext<'b>,
+    supported_compressed_formats: CompressedImageFormats,
+) -> Result<Image, ObjError> {
+    let path = load_context.path().with_file_name(image_path);
+    let extension = ImageType::Extension(
+        path.extension()
+            .and_then(|e| e.to_str())
+            .ok_or(ObjError::InvalidImageFile(path.to_path_buf()))?,
+    );
+    let bytes = load_context.asset_io().load_path(&path).await?;
+    let is_srgb = true;
+    Ok(Image::from_buffer(
+        &bytes,
+        extension,
+        supported_compressed_formats,
+        is_srgb,
+    )?)
+}
+
+async fn load_obj_data<'a, 'b>(
+    mut bytes: &'a [u8],
+    load_context: &'a mut LoadContext<'b>,
+) -> tobj::LoadResult {
+    let options = tobj::GPU_LOAD_OPTIONS;
+    let asset_io = load_context.asset_io();
+    let ctx_path = load_context.path();
+    tobj::load_obj_buf_async(&mut bytes, &options, |p| async move {
+        let asset_path = ctx_path.with_file_name(p);
+        asset_io
+            .load_path(&asset_path)
+            .await
+            .map_or(Err(tobj::LoadError::OpenFileFailed), |bytes| {
+                tobj::load_mtl_buf(&mut bytes.as_slice())
+            })
+    })
+    .await
+}
+
+async fn load_mat_texture<'a, 'b>(
+    texture: &Option<String>,
+    load_context: &'a mut LoadContext<'b>,
+    supported_compressed_formats: CompressedImageFormats,
+) -> Result<Option<Handle<Image>>, ObjError> {
+    if let Some(texture) = texture {
+        let handle = if load_context.has_labeled_asset(texture) {
+            load_context.get_handle(texture)
```

Comment:
I was testing some things and I believe this is causing issues. It does not return the handle of the texture.
Looks like "asset label" isn't the same as `get_handle`'s "asset id".

Reply:
I'll give it a try, do you have an example of snippet / assets that render incorrectly?

Reply:
[cube.zip](https://github.com/AmionSky/bevy_obj/files/11433264/cube.zip)

Reply:
https://github.com/AmionSky/bevy_obj/pull/12/commits/cd4859cf0a79ec2757cd76706ee19bfdf743ad5e should fix it.
I gave a spin to storing only weak handles in the hash map but that didn't really work since casting to strong requires passing the asset collection